railib/api.py

```diff
@@ -664,14 +671,36 @@ def load_json(ctx: Context, database: str, engine: str, relation: str,
                "def insert:%s = load_json[config]" % relation)
     return query(ctx, database, engine, command, inputs=inputs, readonly=False)
 
-
 def query(ctx: Context, database: str, engine: str, command: str,
           inputs: dict = None, readonly: bool = True) -> dict:
     tx = Transaction(database, engine, readonly=readonly)
     return tx.run(ctx, _query_action(command, inputs=inputs))
 
+# Answers if the given transaction state is a terminal state.
+def is_txn_term_state(state: str) -> bool:
+    return state == "COMPLETED" or state == "ABORTED"
+
+def exec(ctx: Context, database: str, engine: str, command: str,
+          inputs: dict = None, readonly: bool = True) -> list:
+    async_result = exec_async(ctx, database, engine, command, readonly=readonly)
+    if isinstance(async_result, list):  # in case of if short-path, return results directly, no need to poll for state
+        return async_result
+
+    rsp = []
+    while True:
+        time.sleep(3)
+        txn = get_transaction(ctx, async_result["id"])
+        if is_txn_term_state(txn["state"]):
+            rsp.append(txn)
+            rsp.append(get_transaction_metadata(ctx, txn["id"]))
+            rsp.append(get_transaction_problems(ctx, txn["id"]))
+            rsp.append(get_transaction_results(ctx, txn["id"]))
+            break
+
+    return rsp
```

Comment:
The other SDKs are currently returning a Dictionary, not a List.

(Well actually, the Julia SDK just changed to a struct called TransactionResponse, which is maybe even better. It was a Dict before that.)

Julia:
```julia
julia> rsp = RAI.exec(ctx, db, eng, """2+2""");

julia> fieldnames(typeof(rsp))
(:transaction, :metadata, :problems, :results)

julia> rsp.transaction
JSON3.Object{Base.CodeUnits{UInt8, String}, Vector{UInt64}} with 3 entries:
  :id                      => "33dccded-4af9-399c-6b6a-bd63bdfe934d"
  :response_format_version => "2.0.3"
  :state                   => "COMPLETED"
```
Go:
https://github.com/RelationalAI/rai-sdk-go/blob/e5f70e84e03511e61047e0baf829f9d78410137c/rai/client.go#L1267
(I would prefer to rename that type in go from TransactionAsyncResult to TransactionResponse, i think.)

Maybe you can use the same approach here? Either make a type, or return a dict?

Reply:
^^ And ideally, the `exec_async()` and `exec()` should return the same structure in both cases. This way the calling code for `exec_async()` can expect the same type to be returned whether or not it hits the fast-path.

Reply:
@NHDaly this turned out to be a lot more complex than I thought it would be. I will merge this and work on what you suggested in another PR.

Reply:
sounds great. Thanks! :)

Reply:
Maybe we can just do a quick PR that changes `exec_async()` to return a single-element List for the async case, and the full List for the sync case? This way the type is the same, and the Transaction resource is always in the same spot: the first element in the list. And the only thing that changes is whether the rest of the data is present or not.

Would that be easy to do in a small PR?

Reply:
how does this work in other SDKs?

I'd prefer something like `exec_async(…).transaction` to `exec_async()[0]`…

Reply:
Yes, that's what i advocated for above, as well, but Sameera wanted to save that for a follow-up work.
https://github.com/RelationalAI/rai-sdk-python/pull/78#discussion_r922243085:~:text=NHDaly-,2%20days%20ago,-The%20other%20SDKs
https://github.com/RelationalAI/rai-sdk-python/pull/78#discussion_r921074142